guides/source/active_storage_overview.md

```diff
@@ -553,6 +553,41 @@ end
 
 NOTE: Since Active Storage relies on polymorphic associations, and [polymorphic associations](./association_basics.html#polymorphic-associations) rely on storing class names in the database, that data must remain synchronized with the class name used by the Ruby code. When renaming classes that use `has_many_attached`, make sure to also update the class names in the `active_storage_attachments.record_type` polymorphic type column of the corresponding rows.
 
+### Callbacks
+
+Active Storage provides callbacks for when attachments are successfully uploaded to the configured service.
+**Before** and **After** callbacks are available on both `has_one_attached` and `has_many_attached` associations.
+The method name will include the attachment name and the callback type.
+_For variants, the suffix `_variant_attached` will be used._
+
+```ruby
+class Message < ApplicationRecord
+  has_one_attached :image do |attachable|
+    attachable.variant :thumb, resize_to_limit: [100, 100]
+  end
+
+  has_many_attached :images do |attachable|
+    attachable.variant :thumb, resize_to_limit: [100, 100]
+  end
+
+  def before_image_attached(image_blob)
```

Comment:
IMHO this feels different from normal AR callbacks. Usually we'd have something like

```
class Message < ApplicationRecord
  has_one_attached :image do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]
  end

  before_image_attached do |image_blob|
    # do stuff
  end
end
```

Minor change, but feels more similar to what we already have with Active Record callbacks. Thoughts?

Reply:
> IMHO this feels different from normal AR callbacks. Usually we'd have something like
> 
> ```ruby
> class Message < ApplicationRecord
>   has_one_attached :image do |attachable|
>     attachable.variant :thumb, resize_to_limit: [100, 100]
>   end
> 
>   before_image_attached do |image_blob|
>     # do stuff
>   end
> end
> ```
> 
> Minor change, but it feels more similar to what we already have with Active Record callbacks. Thoughts?

Thanks @MatheusRich, for sure! I completely agree with you. And this was the original intent. 
Actually that was the first iteration, the problem however was that ar callbacks would not allow us to yield parameters back to it. 

We would have to add that ability to ar callbacks from the ground up and in the end we decided against it and follow this simpler approach.

Reply:
This isn't a fully baked thought, but what about something like:

```
before_image_attached :do_stuff

def do_stuff(blob)
  # do stuff with blob
end
```

Reply:
> This isn't a fully baked thought, but what about something like:
> 
> ```
> before_image_attached :do_stuff
> 
> def do_stuff(blob)
>   # do stuff with blob
> end
> ```

Hi! That is the same proposition on the first comment, please read my [answer to it above](https://github.com/rails/rails/pull/51814#discussion_r1602201440).

Reply:
Great work here! Echoing Robert's earlier message: normal AM callbacks don't allow us to pass an argument. My assumption is that AM callbacks expect to be idempotent, and that all information is available on `self`. However, if you `has_many_attached`, and you want to do some processing on the specific image that was just attached, that info is harder to get on `self` alone. You could query the most recently added attachment, but that could cause race conditions.

I wonder if it would help prevent confusion by differentiating it from the usual ActiveModel Callback pattern. Something like...

```rb
class Message < ApplicationRecord
  # with a lambda
  has_one_attached :image, after_attached: ->(blob) { puts "image attached" } do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]
  end

  # OR, with a method name
  has_one_attached :image, after_attached: :example_after_attached do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]
  end

  def example_after_attached(blob)
  end
end
```

Now it's more like a part of `has_one_attached`'s API rather than trying to mimic the ActiveModel Callback API. You could still send an argument to it like you have here so you get all the same functionality, and the user gets to customize the method name (or use a lambda).

_However_, if folks aren't concerned about this, I think your solution looks good as-is.

Reply:
I actually love your suggestion Matt, it makes a lot of sense to me and feels more explicit (and makes the behavior more readable and deterministic). 

I don't have a strong opinion about leaving it as-is or using the new approach. I'm fine with whichever.

Reply:
That last suggestion feels a lot more like normal ActiveRecord. I like the flexibility of it compared to a magically named method that you have to constantly do `respond_to?` checks for.

One suggestion I was thinking that might be a little more ergonomic than a hash argument in a method that also expects a block, is to add 2 methods to ActiveStorage::Reflection::HasAttachedReflection.

```ruby
class Message < ApplicationRecord
  # with a lambda
  has_one_attached :image do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]

    attachable.after_attached do |blob|
      puts "image attached"
    end
  end

  # OR, with a method name
  has_one_attached :image do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]

    attachable.before_attached :example_before_attached
  end

  def example_before_attached(blob)
  end
end
```

Reply:
It'd also be more performant to have to actually specify at the time of defining the `has_one_attached` that there are callbacks, than to always attempt to run callbacks only to find that none are defined.

Reply:
I went ahead with @mattboldt 's suggestion and made it as named options that support both lambdas and symbols. 
Now it actually looks more _rails-y_  and is more performant.